snowfall/decoding/lm_rescore.py

```diff
@@ -0,0 +1,268 @@
+    Returns:
+      An FsaVec representing the best decoding path for each sequence
+      in the lattice.
+    '''
+    device = lats.device
+
+    assert len(lats.shape) == 3
+    assert hasattr(lats, 'aux_labels')
+    assert hasattr(lats, 'lm_scores')
+
+    assert G.shape == (1, None, None)
+    assert G.device == device
+    assert hasattr(G, 'aux_labels') is False
+
+    # First, extract `num_paths` paths for each sequence.
+    # paths is a k2.RaggedInt with axes [seq][path][arc_pos]
+    paths = k2.random_paths(lats, num_paths=num_paths, use_double_scores=True)
+
+    # word_seqs is a k2.RaggedInt sharing the same shape as `paths`
+    # but it contains word IDs. Note that it also contains 0s and -1s.
+    # The last entry in each sublist is -1.
+    word_seqs = k2.index(lats.aux_labels, paths)
+
+    # Remove epsilons and -1 from word_seqs
+    word_seqs = k2.ragged.remove_values_leq(word_seqs, 0)
+
+    # Remove repeated sequences to avoid redundant computation later.
+    #
+    # unique_word_seqs is still a k2.RaggedInt with 3 axes [seq][path][word]
+    # except that there are no repeated paths with the same word_seq
+    # within a seq.
+    #
+    # num_repeats is also a k2.RaggedInt with 2 axes containing the
+    # multiplicities of each path.
+    # num_repeats.num_elements() == unique_word_seqs.num_elements()
+    #
+    # Since k2.ragged.unique_sequences will reorder paths within a seq,
+    # `new2old` is a 1-D torch.Tensor mapping from the output path index
+    # to the input path index.
+    # new2old.numel() == unique_word_seqs.num_elements()
+    unique_word_seqs, num_repeats, new2old = k2.ragged.unique_sequences(
+        word_seqs, need_num_repeats=True, need_new2old_indexes=True)
+
+    seq_to_path_shape = k2.ragged.get_layer(unique_word_seqs.shape(), 0)
+
+    # path_to_seq_map is a 1-D torch.Tensor.
+    # path_to_seq_map[i] is the seq to which the i-th path
+    # belongs.
+    path_to_seq_map = seq_to_path_shape.row_ids(1)
+
+    # Remove the seq axis.
+    # Now unique_word_seqs has only two axes [path][word]
+    unique_word_seqs = k2.ragged.remove_axis(unique_word_seqs, 0)
+
+    # word_fsas is an FsaVec with axes [path][state][arc]
+    word_fsas = k2.linear_fsa(unique_word_seqs)
+
+    word_fsas_with_epsilon_loops = k2.add_epsilon_self_loops(word_fsas)
+
+    am_scores = compute_am_scores(lats, word_fsas_with_epsilon_loops,
+                                  path_to_seq_map)
+
+    # Now compute lm_scores
+    b_to_a_map = torch.zeros_like(path_to_seq_map)
+    lm_path_lats = k2.intersect_device(G,
+                                       word_fsas_with_epsilon_loops,
+                                       b_to_a_map=b_to_a_map,
+                                       sorted_match_a=True)
+    lm_path_lats = k2.top_sort(k2.connect(lm_path_lats.to('cpu'))).to(device)
+    lm_scores = lm_path_lats.get_tot_scores(True, True)
+
+    tot_scores = am_scores + lm_scores
```

Comment:
doesn't this way of rescoring remove the "graph scores" that come from e.g. multiple entries with distinct probabilities in the lexicon?

Reply:
I believe we are currently using a lexicon with log-prob 0 on all arcs, so the lexicon does not contribute probabilities to the final rescoring.

The current approach assumes that LM scores come exclusively from the language model.

Reply:
What I am currently doing is

- Get an n-best list.  (Note that the `best` in `n-best` is not the same one as in Kaldi. This list is sampled from a lattice)
- For each list (i.e., for each path), we remove the scores of very arc that come from the language model (i.e., HLG).
- Get the acoustic scores of each path
- Get the LM scores of each path
- Get the total scores of each path, with is am_scores + lm_scores
- The path with the largest total scores is the best path, which is used as the decoding output

Reply:
What I am currently doing is

- Get an n-best list.  (Note that the `best` in `n-best` is not the same one as in Kaldi. This list is sampled from a lattice)
- For each list (i.e., for each path), we remove the scores of very arc that come from the language model (i.e., HLG).
- Get the acoustic scores of each path
- Get the LM scores of each path
- Get the total scores of each path, with is am_scores + lm_scores
- The path with the largest total scores is the best path, which is used as the decoding output

Reply:
Sorry I missed this reply. My question was this: in Kaldi, HCLG arc scores actually consisted of two types of scores: the LM scores, and "graph" scores, which came from pronunciation probabilities in lexiconp. AFAIK rescoring in Kaldi always first re-computed the LM scores with the original G and subtracted them from the LM scores in the lattice, so that the "graph" scores were preserved, and only then added the rescoring LM scores on top of that.

I am not sure if our HLG supports multiple pronunciations and their probabilities (or do we even need to do it anymore), but it's just something that came to my mind when reviewing.